### DISCUSSION TITLE
How to pass gradients to `backward()`

### DISCUSSION THREAD

**Discussion by JayMan91** [Lightning-AI/pytorch-lightning#10257] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
In my experiment, I have a loss function, which is not defined by an expression. But I have a formulation of the gradient (formally a subgradient) so I have to pass the gradient manually. In pytorch, I implement it in the following way and it is working fine.

```python
self.optimizer.zero_grad()
y_hat = self.model(x_train)
grad =  compute_grad(y_hat, y)
y_hat.backward(gradient=grad)
self.optimizer.step()
```
Would the following be a correct implementation in lightning?
```python
def training_step(self, batch, batch_idx):
        opt = self.optimizers()
        x,y = batch
        y_hat =  self(x)
        grad =  compute_grad(y_hat, y)
        opt.zero_grad()
        y_hat.backward(gradient= grad)
        opt.step()
```

**Comment by akihironitta** [comment, marked as answer]
@JayMan91 Haven't tried, but you could probably try manual optimization.

```python
def __init__(...):
    ...
    self.automatic_optimization = False  # use manual optimization
    ...

def training_step(...):
    ...
    self.manual_backward(y_hat, gradient=grad)
    ...
```

manual optimization: https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#manual-optimization
`LightningModule.manual_backward`: https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#manual-backward

**Reply by JayMan91** [reply]
Thanks for your reply. In this case, we have to write optimizer.step() and optimizer.zero_grad(), right?

**Reply by akihironitta** [reply]
Yes, you need to call them manually in your training step.
